crates/typst/src/layout/align.rs

```diff
@@ -445,6 +476,150 @@ cast! {
     }
 }
 
+/// An internal representation that combines horizontal or vertical alignments. The
+/// allowed alignment positions are designated by the type parameter `H` and `V`.
+///
+/// This is not user-visible, but an internal type to impose type safety. For example,
+/// `SpecificAlignment<HAlignment, OuterVAlignment>` does not allow vertical alignment
+/// position "center", because `V = OuterVAlignment` doesn't have it.
+#[derive(Debug, Copy, Clone, Eq, PartialEq, Hash)]
+pub enum SpecificAlignment<H, V>
+where
+    H: Into<HAlignment> + Copy,
+    V: Into<VAlignment> + Copy,
+    HAlignment: Into<H>,
+    VAlignment: Into<V>,
+{
+    H(H),
+    V(V),
+    Both(H, V),
+}
+
+impl<H, V> SpecificAlignment<H, V>
+where
+    H: Into<HAlignment> + Copy,
+    V: Into<VAlignment> + Copy,
+    HAlignment: Into<H>,
+    VAlignment: Into<V>,
+{
+    /// The horizontal component.
+    pub const fn x(self) -> Option<H> {
+        match self {
+            Self::H(x) | Self::Both(x, _) => Some(x),
+            Self::V(_) => None,
+        }
+    }
+
+    /// The vertical component.
+    pub const fn y(self) -> Option<V> {
+        match self {
+            Self::V(y) | Self::Both(_, y) => Some(y),
+            Self::H(_) => None,
+        }
+    }
+}
+
+impl Add<OuterVAlignment> for HAlignment {
+    type Output = SpecificAlignment<HAlignment, OuterVAlignment>;
+
+    fn add(self, rhs: OuterVAlignment) -> Self::Output {
+        SpecificAlignment::Both(self, rhs)
+    }
+}
+
+impl Add<HAlignment> for OuterVAlignment {
+    type Output = SpecificAlignment<HAlignment, OuterVAlignment>;
+
+    fn add(self, rhs: HAlignment) -> Self::Output {
+        SpecificAlignment::Both(rhs, self)
+    }
+}
+
```

Comment:
I don't think we will use this anywhere in the Rust code, since you can just write `SpecificAlignment::Both(first, second)` instead (which is much clearer). Users won't use this either, since variables in Typst can't have type OuterHAlignment nor OuterVAlignment, given that a user can only have a variable of type Alignment (which may correspond to HAlignment or VAlignment, or both, under the hood).

```suggestion
```